beacon_node/store/src/migrate.rs

```diff
@@ -0,0 +1,138 @@
+            // This migrator is only used for testing, so we just log to stderr without a logger.
+            eprintln!("Migration error: {:?}", e);
+        }
+    }
+}
+
+/// Migrator that runs a background thread to migrate state from the hot to the cold database.
+pub struct BackgroundMigrator<E: EthSpec> {
+    db: Arc<RwLock<DiskStore>>,
+    tx_thread: Mutex<(
+        mpsc::Sender<(Hash256, BeaconState<E>)>,
+        thread::JoinHandle<()>,
+    )>,
+}
+
+impl<E: EthSpec> Migrate<DiskStore, E> for BackgroundMigrator<E> {
+    fn new(db: Arc<RwLock<DiskStore>>) -> Self {
+        let tx_thread = Mutex::new(Self::spawn_thread(db.clone()));
+        Self { db, tx_thread }
+    }
+
+    /// Perform the freezing operation on the database,
+    fn freeze_to_state(
+        &self,
+        finalized_state_root: Hash256,
+        finalized_state: BeaconState<E>,
+        max_finality_distance: u64,
+    ) {
+        if !self.needs_migration(finalized_state.slot, max_finality_distance) {
+            return;
+        }
+
+        let (ref mut tx, ref mut thread) = *self.tx_thread.lock();
+
+        if let Err(tx_err) = tx.send((finalized_state_root, finalized_state)) {
+            let (new_tx, new_thread) = Self::spawn_thread(self.db.clone());
+
+            drop(mem::replace(tx, new_tx));
+            let old_thread = mem::replace(thread, new_thread);
+
+            // Join the old thread, which will probably have panicked, or may have
+            // halted normally just now as a result of us dropping the old `mpsc::Sender`.
+            if let Err(thread_err) = old_thread.join() {
+                warn!(
+                    self.db.read().log,
+                    "Migration thread died, so it was restarted";
+                    "reason" => format!("{:?}", thread_err)
+                );
+            }
+
+            // Retry at most once, we could recurse but that would risk overflowing the stack.
+            let _ = tx.send(tx_err.0);
+        }
+    }
+}
+
+impl<E: EthSpec> BackgroundMigrator<E> {
+    /// Return true if a migration needs to be performed, given a new `finalized_slot`.
+    fn needs_migration(&self, finalized_slot: Slot, max_finality_distance: u64) -> bool {
+        let finality_distance = finalized_slot - self.db.read().get_split_slot();
+        finality_distance > max_finality_distance
+    }
+
+    /// Spawn a new child thread to run the migration process.
+    ///
+    /// Return a channel handle for sending new finalized states to the thread.
+    fn spawn_thread(
+        db: Arc<RwLock<DiskStore>>,
+    ) -> (
+        mpsc::Sender<(Hash256, BeaconState<E>)>,
+        thread::JoinHandle<()>,
+    ) {
+        let (tx, rx) = mpsc::channel();
+        let thread = thread::spawn(move || {
+            while let Ok((state_root, state)) = rx.recv() {
+                if let Err(e) = DiskStore::freeze_to_state(db.clone(), state_root, &state) {
+                    warn!(db.read().log, "Migration error: {:#?}", e);
```

Comment:
A log like this might be more consistent with the rest of the binary:

```rust
warn!(
  db.read().log,
  "Database migration failed";
  "error" => format!("{:#?}", e)
)
```